specs/phase0/beacon-chain.md

```diff
@@ -603,18 +607,50 @@ def bytes_to_uint64(data: bytes) -> uint64:
 
 #### BLS Signatures
 
+##### IETF standard
+
 Eth2 makes use of BLS signatures as specified in the [IETF draft BLS specification draft-irtf-cfrg-bls-signature-03](https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-03). Specifically, eth2 uses the `BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_` ciphersuite which implements the following interfaces:
 
-- `def Sign(SK: int, message: Bytes) -> BLSSignature`
-- `def Verify(PK: BLSPubkey, message: Bytes, signature: BLSSignature) -> bool`
+- `def Sign(SK: int, message: bytes) -> BLSSignature`
 - `def Aggregate(signatures: Sequence[BLSSignature]) -> BLSSignature`
-- `def FastAggregateVerify(PKs: Sequence[BLSPubkey], message: Bytes, signature: BLSSignature) -> bool`
-- `def AggregateVerify(PKs: Sequence[BLSPubkey], messages: Sequence[Bytes], signature: BLSSignature) -> bool`
+- `def _Verify(PK: BLSPubkey, message: bytes, signature: BLSSignature) -> bool`
+- `def _AggregateVerify(PKs: Sequence[BLSPubkey], messages: Sequence[bytes], signature: BLSSignature) -> bool`
+- `def _FastAggregateVerify(PKs: Sequence[BLSPubkey], message: bytes, signature: BLSSignature) -> bool`
 
 Within these specifications, BLS signatures are treated as a module for notational clarity, thus to verify a signature `bls.Verify(...)` is used.
 
 *Note*: The non-standard configuration of the BLS and hash to curve specs is temporary and will be resolved once IETF releases BLS spec draft 3.
 
+##### Eth2 BLS wrappers
+
+We use the following Eth2-specific BLS wrappers to be compatible with IETF BLS standard.
+
+```python
+def is_zero_sk(PK: int) -> bool:
+    return PK == G1_INFINATY_POINT
+```
+
+```python
+def Eth2Verify(PK: BLSPubkey, message: bytes, signature: BLSSignature) -> bool:
+    if is_zero_sk(PK) and signature == G2_INFINATY_POINT:
+        return True
+    else:
+        return bls._Verify(PK, message, signature)
+```
+
+```python
+def Eth2AggregateVerify(PKs: Sequence[BLSPubkey], messages: Sequence[bytes], signature: BLSSignature) -> bool:
+    PKs = [PK for PK in PKs if not is_zero_sk(PK)]
+    messages = [message for PK, message in zip(PKs, messages) if not is_zero_sk(PK)]
+    return bls._AggregateVerify(PKs, messages, signature)
```

Comment:
I think this is unsafe. An attacker can put zero elements in different locations of `PKs` and `messages`

Reply:
i'm not sure what you mean by "zero elements in different locations of `PKs` and `messages`" -- zero messages?

but either way this implementation will drop messages so we should do the filtering and selecting together...

e.g. as written:
```python
# given:
PKs = [p1, 0, p3]
messages = [m1, m2, m3]

# inside function:
#   after first line:
PKs = [p1, p3]
#   after second line:
messages = [m1, m2]
```

alternative:
```python
# inside function...
# possibly `assert` ourselves that `len(PKs) == len(messages)`
pubkeys_and_messages = [(pubkey, message) for pubkey, message in zip(PKs, messages) if not is_zero_sk(pubkey)]
pubkeys, messages = list(zip(*pubkeys_and_messages))
return bls._AggregateVerify(pubkeys, messages, signature)
```

Reply:
yeah we can definitely unroll that to something more readable :)